crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,532 @@
+            _ => return None,
+        }
+    }
+    Some(lines)
+}
+
+#[derive(Debug)]
+struct LineWithJustAComment(TextRange);
+
+impl LineWithJustAComment {
+    fn new(comment_range: TextRange, total_dunder_all_range: TextRange) -> Self {
+        Self(comment_range + total_dunder_all_range.start())
+    }
+}
+
+#[derive(Debug)]
+struct LineWithItems {
+    items: Vec<(String, TextRange)>,
+    comment_range: Option<TextRange>,
+}
+
+impl LineWithItems {
+    fn new(
+        items: &[(String, TextRange)],
+        comment_range: Option<TextRange>,
+        total_dunder_all_range: TextRange,
+    ) -> Self {
+        assert!(
+            !items.is_empty(),
+            "Use the 'JustAComment' variant to represent lines with 0 items"
+        );
+        let offset = total_dunder_all_range.start();
+        Self {
+            items: items
+                .iter()
+                .map(|(s, r)| (s.to_owned(), r + offset))
+                .collect(),
+            comment_range: comment_range.map(|c| c + offset),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum DunderAllLine {
+    JustAComment(LineWithJustAComment),
+    OneOrMoreItems(LineWithItems),
+}
+
+fn collect_dunder_all_items(lines: &[DunderAllLine]) -> Vec<DunderAllItem> {
+    // Given data on each line in `__all__`, group lines together into "items".
+    // Each item contains exactly one element,
+    // but might contain multiple comments attached to that element
+    // that must move with the element when `__all__` is sorted.
+    let mut all_items = vec![];
+    let mut this_range = None;
+    for line in lines {
+        match line {
+            DunderAllLine::JustAComment(LineWithJustAComment(comment_range)) => {
+                this_range = Some(*comment_range);
+            }
+            DunderAllLine::OneOrMoreItems(LineWithItems {
+                items,
+                comment_range,
+            }) => {
+                let [(first_val, first_range), rest @ ..] = items.as_slice() else {
+                    unreachable!(
+                        "LineWithItems::new() should uphold the invariant that this list is always non-empty"
+                    )
+                };
+                let range = this_range.map_or(*first_range, |r| {
+                    TextRange::new(r.start(), first_range.end())
+                });
+                all_items.push(DunderAllItem {
+                    value: first_val.clone(),
+                    original_index: all_items.len(),
+                    range,
+                    additional_comments: *comment_range,
+                });
+                this_range = None;
+                for (value, range) in rest {
+                    all_items.push(DunderAllItem {
+                        value: value.clone(),
+                        original_index: all_items.len(),
+                        range: *range,
+                        additional_comments: None,
+                    });
+                }
```

Comment:
We can use `Vec::extend` here to let `Vec` know that we intend to push multiple items. It can help `Vec` to regrow the vector only once


```suggestion
                all_items.extend(rest.map(|(value, range)| DunderAllItem {
                    value,
                    original_index: all_items.len(),
                    range,
                    additional_comments: None,
                }));
```

Reply:
(For posterity: we tried this out in a pair-programming session, but the borrow checker was very angry, so we reverted it)